roachtest: engine/switch/nodes=3 failed

[(roachtest).engine/switch/nodes=3 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1775192&tab=buildLog) on [master@3e2334ded18d2f18c08f2f912be085df563ca6ff](https://github.com/cockroachdb/cockroach/commits/3e2334ded18d2f18c08f2f912be085df563ca6ff):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/engine/switch/nodes=3/run_1
	cluster.go:2368,engine_switch.go:126,engine_switch.go:140,test_runner.go:741: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2356
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2364
		    main.registerEngineSwitch.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/engine_switch.go:126
		    main.registerEngineSwitch.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/engine_switch.go:140
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:741
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - unexpected node event: 1: dead

	cluster.go:1410,context.go:135,cluster.go:1399,test_runner.go:778: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1775192-1582963470-12-n4cpu4 --oneshot --ignore-empty-nodes: exit status 1 1: dead
		2: 4471
		4: skipped
		3: 4490
		Error:  1: dead

```

<details><summary>More</summary><p>


Artifacts: [/engine/switch/nodes=3](https://teamcity.cockroachdb.com/viewLog.html?buildId=1775192&tab=artifacts#/engine/switch/nodes=3)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Aengine%2Fswitch%2Fnodes%3D3.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).engine/switch/nodes=3 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1780216&tab=buildLog) on [master@5f9a71adb995837bcff27b9456188018434be4b8](https://github.com/cockroachdb/cockroach/commits/5f9a71adb995837bcff27b9456188018434be4b8):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200303-1780216/engine/switch/nodes=3/run_1
	cluster.go:2368,engine_switch.go:126,engine_switch.go:140,test_runner.go:741: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2356
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2364
		    main.registerEngineSwitch.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/engine_switch.go:126
		    main.registerEngineSwitch.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/engine_switch.go:140
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:741
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - unexpected node event: 1: dead

	cluster.go:1410,context.go:135,cluster.go:1399,test_runner.go:778: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1780216-1583223197-08-n4cpu4 --oneshot --ignore-empty-nodes: exit status 1 4: skipped
		3: 4527
		2: 4309
		1: dead
		Error:  1: dead

```

<details><summary>More</summary><p>


Artifacts: [/engine/switch/nodes=3](https://teamcity.cockroachdb.com/viewLog.html?buildId=1780216&tab=artifacts#/engine/switch/nodes=3)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Aengine%2Fswitch%2Fnodes%3D3.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).engine/switch/nodes=3 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1786843&tab=buildLog) on [master@c9b189b3a2171c8da78c75fe80dd02933e18a5e5](https://github.com/cockroachdb/cockroach/commits/c9b189b3a2171c8da78c75fe80dd02933e18a5e5):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200305-1786843/engine/switch/nodes=3/run_1
	cluster.go:2368,engine_switch.go:126,engine_switch.go:140,test_runner.go:741: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2356
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2364
		    main.registerEngineSwitch.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/engine_switch.go:126
		    main.registerEngineSwitch.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/engine_switch.go:140
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:741
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - unexpected node event: 1: dead

	cluster.go:1410,context.go:135,cluster.go:1399,test_runner.go:778: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1786843-1583394619-14-n4cpu4 --oneshot --ignore-empty-nodes: exit status 1 4: skipped
		1: dead
		2: 4432
		3: 4395
		Error:  1: dead

```

<details><summary>More</summary><p>


Artifacts: [/engine/switch/nodes=3](https://teamcity.cockroachdb.com/viewLog.html?buildId=1786843&tab=artifacts#/engine/switch/nodes=3)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Aengine%2Fswitch%2Fnodes%3D3.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/storage/engine_test.go
+++ b/pkg/storage/engine_test.go
@@ -419,6 +419,15 @@ func TestEnginePutGetDelete(t *testing.T) {
 	}
 }
 
+func addMergeTimestamp(t *testing.T, data []byte, ts int64) []byte {
+	var v enginepb.MVCCMetadata
+	if err := protoutil.Unmarshal(data, &v); err != nil {
+		t.Fatal(err)
+	}
+	v.MergeTimestamp = &hlc.LegacyTimestamp{WallTime: ts}
+	return mustMarshal(&v)
+}
+
 // TestEngineMerge tests that the passing through of engine merge operations
 // to the goMerge function works as expected. The semantics are tested more
 // exhaustively in the merge tests themselves.
@@ -447,28 +456,28 @@ func TestEngineMerge(t *testing.T) {
 				{
 					mvccKey("timeseriesmerged"),
 					[][]byte{
-						timeSeriesRow(testtime, 1000, []tsSample{
+						addMergeTimestamp(t, timeSeriesRow(testtime, 1000, []tsSample{
 							{1, 1, 5, 5, 5},
-						}...),
+						}...), 27),
 						timeSeriesRow(testtime, 1000, []tsSample{
 							{2, 1, 5, 5, 5},
 							{1, 2, 10, 7, 3},
 						}...),
-						timeSeriesRow(testtime, 1000, []tsSample{
+						addMergeTimestamp(t, timeSeriesRow(testtime, 1000, []tsSample{
 							{10, 1, 5, 5, 5},
-						}...),
+						}...), 53),
 						timeSeriesRow(testtime, 1000, []tsSample{
 							{5, 1, 5, 5, 5},
 							{3, 1, 5, 5, 5},
 						}...),
 					},
-					timeSeriesRow(testtime, 1000, []tsSample{
+					addMergeTimestamp(t, timeSeriesRow(testtime, 1000, []tsSample{
 						{1, 2, 10, 7, 3},
 						{2, 1, 5, 5, 5},
 						{3, 1, 5, 5, 5},
 						{5, 1, 5, 5, 5},
 						{10, 1, 5, 5, 5},
-					}...),
+					}...), 27),
 				},
 			}
 			for _, tc := range testcases {

--- a/pkg/storage/pebble_merge.go
+++ b/pkg/storage/pebble_merge.go
@@ -15,6 +15,7 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 	"github.com/cockroachdb/errors"
 	"github.com/gogo/protobuf/proto"
@@ -169,6 +170,7 @@ func ensureColumnar(ts *roachpb.InternalTimeSeriesData) {
 type MVCCValueMerger struct {
 	timeSeriesOps []roachpb.InternalTimeSeriesData
 	rawByteOps    [][]byte
+	oldestMergeTS hlc.LegacyTimestamp
 	oldToNew      bool
 
 	// Used to avoid heap allocations when passing pointer to `Unmarshal()`.
@@ -218,6 +220,11 @@ func (t *MVCCValueMerger) deserializeMVCCValueAndAppend(value []byte) error {
 			return errors.Errorf("inconsistent value types for non-timeseries merge")
 		}
 		t.rawByteOps = append(t.rawByteOps, t.meta.RawBytes[mvccHeaderSize:])
+	}
+	// Save the timestamp of the oldest value since that is consistent with the
+	// behavior of the C++ DBMergeOperator.
+	if t.meta.MergeTimestamp != nil && (t.oldestMergeTS == hlc.LegacyTimestamp{} || !t.oldToNew) {
+		t.oldestMergeTS = *t.meta.MergeTimestamp
 	}
 	return nil
 }
@@ -307,6 +314,9 @@ func (t *MVCCValueMerger) Finish() ([]byte, error) {
 		return nil, err
 	}
 	var meta enginepb.MVCCMetadata
+	if !(t.oldestMergeTS == hlc.LegacyTimestamp{}) {
+		meta.MergeTimestamp = &t.oldestMergeTS
+	}
 	tsTag := byte(roachpb.ValueType_TIMESERIES)
 	header := make([]byte, mvccHeaderSize)
 	header[mvccTagPos] = tsTag

